File: rust/parity/specs/llm_output_sanitizer.py
"""Parity spec for PiLLMOutputSanitizer.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiLLMOutputSanitizer"

_mod = load_py_agent("pi_llm_output_sanitizer.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiLLMOutputSanitizer()
    out = agent.sanitize_llm_output(_mod.LLMOutputSanitizerInput(**data))
    return out.model_dump()


SAMPLES = [
    # 1. Clean, passing output -> risk 0, CLEAN
    {"input": {"raw_output": "The weather today is sunny and pleasant."}},
    # 2. Single system-prompt trigger -> risk 35.0, still secure (< 40)
    {"input": {"raw_output": "You are a helpful assistant for booking flights."}},
    # 3. Two system-prompt triggers -> risk 70.0 -> COMPROMISED
    {"input": {"raw_output": "The system prompt says to ignore previous instructions."}},
    # 4. Credential leak with literal api_key -> redacted + COMPROMISED
    {"input": {"raw_output": "Use api_key=sk-12345 to authenticate the request."}},
    # 5. Bearer token leak (no api_key, so no substring scrub) -> COMPROMISED
    {"input": {"raw_output": "Authorization: Bearer eyJhbGciOiJI token granted."}},
    # 6. Combined triggers + credential, exceeds cap -> risk_score clamps to 100.0
    {"input": {"raw_output": (
        "you are a helpful assistant. system prompt: internal guidelines. "
        "ignore previous instructions. api_key=abc aws_access here."
    )}},
    # 7. Empty raw_output edge case -> CLEAN, empty sanitized
    {"input": {"raw_output": ""}},
    # 8. Uppercase API_KEY: matched by raw.lower() check but .replace('api_key') is
    #    case-sensitive so the original-case text is NOT scrubbed (parity gotcha)
    {"input": {"raw_output": "Your API_KEY is exposed here.",
               "system_prompt_reference": "v2-guidelines"}},
]

File: rust/crates/pi-agents/src/agents/llm_output_sanitizer.rs
//! Port of `pi_micro_agents/pi_llm_output_sanitizer.py`.
//!
//! Sanitizes raw LLM output by detecting system-prompt-exposure triggers and
//! leaked credentials, computing a risk score, and scrubbing obvious secrets.
//! Behaviour is a line-for-line mirror of the Python original.

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub raw_output: String,
    #[serde(default = "default_system_prompt_reference")]
    pub system_prompt_reference: String,
}

fn default_system_prompt_reference() -> String {
    "".to_string()
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub sanitized_output: String,
    pub risk_score: f64,
    pub detected_leaks: Vec<String>,
    pub status: String,
}

pub fn sanitize(input: &Input) -> Output {
    let raw = &input.raw_output;
    let mut leaks: Vec<String> = Vec::new();
    let mut risk_score: f64 = 0.0;
    let mut sanitized: String = raw.clone();

    // Lowercased view used for the membership checks (Python: raw.lower()).
    let raw_lower = raw.to_lowercase();

    // Check for system instruction leaks
    let system_triggers = [
        "you are a helpful assistant",
        "system prompt",
        "internal guidelines",
        "ignore previous instructions",
    ];
    for trigger in system_triggers {
        if raw_lower.contains(trigger) {
            leaks.push(format!(
                "Detected potential system prompt exposure trigger: '{trigger}'"
            ));
            risk_score += 35.0;
        }
    }

    // Check for standard credential leaks
    let credential_tokens = ["api_key", "bearer ", "aws_access"];
    if credential_tokens.iter().any(|tok| raw_lower.contains(tok)) {
        leaks.push("Detected leaked authorization token in output.".to_string());
        risk_score += 45.0;
        // Simple scrub
        sanitized = sanitized.replace("api_key", "[REDACTED_API_KEY]");
    }

    risk_score = risk_score.min(100.0);
    let is_secure = risk_score < 40.0;
    let status = if is_secure { "CLEAN" } else { "COMPROMISED" }.to_string();

    Output {
        is_secure,
        sanitized_output: sanitized,
        risk_score,
        detected_leaks: leaks,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = sanitize(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(raw: &str) -> Output {
        sanitize(&Input {
            raw_output: raw.into(),
            system_prompt_reference: "".into(),
        })
    }

    #[test]
    fn clean_output_passes() {
        let o = run("The weather today is sunny and pleasant.");
        assert!(o.is_secure);
        assert_eq!(o.status, "CLEAN");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.detected_leaks.is_empty());
    }

    #[test]
    fn system_prompt_trigger_flagged() {
        // single 35.0 trigger -> still < 40.0 so secure
        let o = run("You are a helpful assistant designed to help.");
        assert!(o.is_secure);
        assert_eq!(o.risk_score, 35.0);
        assert_eq!(o.detected_leaks.len(), 1);
        assert_eq!(o.status, "CLEAN");
    }

    #[test]
    fn credential_leak_compromised_and_redacted() {
        let o = run("Here is your api_key for access.");
        assert!(!o.is_secure);
        assert_eq!(o.risk_score, 45.0);
        assert_eq!(o.status, "COMPROMISED");
        assert!(o.sanitized_output.contains("[REDACTED_API_KEY]"));
    }

    #[test]
    fn risk_score_capped_at_100() {
        // 4 triggers (140) + credential (45) -> capped at 100
        let o = run(
            "you are a helpful assistant. system prompt: internal guidelines. \
ignore previous instructions. api_key=abc",
        );
        assert_eq!(o.risk_score, 100.0);
        assert!(!o.is_secure);
    }
}
